# Spam Detector

In [17]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [20]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

Prediction: Random forest classifier will perform better than the Logistic Regression

Random Forests can capture non-linear relationships in the data through its collection of decision trees, whereas Logistic Regression assumes a linear relationship between features and the outcome. If the patterns in the data are complex or non-linear, Random Forests could have an advantage.

## Split the Data into Training and Testing Sets

In [191]:
# Create the labels set `y` and features DataFrame `X`
y = data["spam"]

X = data.copy()
X = X.drop(columns = "spam")

X.head()


word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
0             0.00            0.00  ...                   0.0         0.00   
1             0.00            0.94  ...                   0.0         0.00   
2             0.64            0.25  ...                   0.0         0.01   
3             0.31            0.63  ...                   0.0         0.00   
4             0.31            0.63  ...                   0.0         0.00   

   char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0        0.000          0.0        0.778        0.000        0.000   
1        0.132          0.0        0.372        0.180        0.048   
2        0.143          0.0        0.276        0.184        0.010   
3        0.137          0.0        0.137        0.000        0.000   
4        0.135          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  
0                       278  
1                      1028  
2                      2259  
3                       191  
4                       191  

[5 rows x 57 columns]

In [34]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
data["spam"].value_counts()

spam
0    2788
1    1813
Name: count, dtype: int64

In [120]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y,)

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [122]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
scaler = StandardScaler()


In [124]:
# Fit the Standard Scaler with the training data
scaler = StandardScaler().fit(X_train)


In [126]:
# Scale the training data
X_train_scaled = scaler.transform(X_train)
X_train_scaled


array([[-0.33741381,  0.2683929 ,  0.50657099, ..., -0.07488932,
        -0.1491885 , -0.18661162],
       [-0.33741381, -0.16616632, -0.56213062, ..., -0.11438826,
        -0.21559166, -0.3739702 ],
       [ 0.57801455, -0.16616632, -0.56213062, ..., -0.10810675,
        -0.16816083, -0.1077238 ],
       ...,
       [-0.33741381, -0.16616632, -0.56213062, ..., -0.03822949,
        -0.17290392, -0.35589174],
       [-0.33741381, -0.05957632,  0.54689936, ...,  0.02859575,
         0.14013953,  0.52666577],
       [-0.33741381, -0.16616632, -0.56213062, ..., -0.13802374,
        -0.24405015, -0.43971005]])

In [128]:
# Transform the test dataset based on the fit from the training dataset
X_test_scaled = scaler.transform(X_test)
X_test_scaled

array([[-0.33741381,  1.17030828, -0.56213062, ..., -0.05661264,
        -0.19661933, -0.36410922],
       [-0.33741381, -0.16616632, -0.56213062, ..., -0.07719609,
        -0.14444542, -0.23756   ],
       [ 1.49344292, -0.16616632,  0.12345155, ...,  0.11721106,
         1.29745165,  1.77900994],
       ...,
       [-0.05331535, -0.16616632, -0.17901117, ..., -0.09376932,
        -0.16341775,  0.3343767 ],
       [-0.33741381, -0.16616632, -0.56213062, ..., -0.03865535,
        -0.11124384, -0.39369215],
       [-0.33741381, -0.16616632, -0.56213062, ..., -0.04994076,
        -0.22507782,  9.24377009]])

In [130]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression

logistic_regression_model = LogisticRegression(random_state=1)

lr_model = logistic_regression_model.fit(X_train_scaled, y_train)

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [145]:
# Make and save testing predictions with the saved logistic regression model using the test data
testing_predections = lr_model.predict(X_test_scaled)

# Review the predictions
testing_predections

array([1, 0, 1, ..., 0, 0, 1], dtype=int64)

In [147]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy_score(y_test, testing_predections)

0.9218071242397915

## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [175]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier

rf_model = RandomForestClassifier(random_state=1).fit(X_train_scaled, y_train)

In [181]:
# Make and save testing predictions with the saved logistic regression model using the test data
predictions = rf_model.predict(X_test_scaled)

# Review the predictions
predictions

array([1, 0, 1, ..., 0, 0, 0], dtype=int64)

In [188]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
acc_score = accuracy_score(y_test, predictions)
acc_score

0.9530842745438749

## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

As I predicted The Random Forest Classifier Model out performed the logistic regression model 95.31% accuracy to 92.18%.

I feel that there were numerous variables in the data set that had more than just linear relationships to spam which was better identified and predicted through the multiple decision trees of the Random Forest Classifier Model.